Split Signature.__or__ into subclasses' __or__

treszkai opened the pull request:
## Description
When task chaining with the bar operator (`X | Y | Z`) was introduced in 2016, the following comment was put in `Signature.__or__`:

> These could be implemented in each individual class, I'm sure, but for now we have this.

I saw this as an invitation to move the quite complex branching into `_chain.__or__`, `group.__or__` and `_chord.__or__`.

The commits of micro-refactoring steps are left as is for easier review.

thedrow commented:
OMG I've been meaning to refractor this for ages. 😳
Thank you! Will review soon.

thedrow commented:
It looks like the PR is perfect and ready for merge but our integration test suite is currently broken.
@treszkai Did you try running it locally? Also, could you please rebase so that the build will be green?

#7149 should resolve all our problems with the integration suite.

@celery/core-developers Should we merge this now and fix whatever problems which arise (if any) after the integration tests run again?

treszkai commented:
@thedrow I rebased now on `master`.

I tried running [the dockerized `make test`](https://docs.celeryproject.org/en/master/contributing.html#developing-and-testing-with-docker), which exits immediately with errors (`rm: missing operand` and `make: python: No such file or directory`).

Running [the tests locally](https://docs.celeryproject.org/en/master/contributing.html#running-the-unit-test-suite) also resulted in immediate errors (`'flaky' not found in `markers` configuration option` in the integration tests and `ModuleNotFoundError: No module named 'dns'` in the unit tests), after installing packages from both requirements files.

The canvas-related tests passed last time in the GitHub workflows.

auvipy commented:
The integration test is not active in CI yet. I returned home today and will try to check the relevant issues

treszkai commented:
@auvipy The decrease in code coverage is due to missing both a `(chord | group) == chain` and `(chord | chain) == chain` in `_chord.__or__` [(L1387)](https://github.com/celery/celery/pull/7135/files#diff-3a80ff45da16a11b96e26a63973d7d490187a68ddc1949e2dfd7fd090b208841R1387)

L414 in the original was _covered_ (but not tested) as `Signature | group` (in `test_group_to_chord`, `test_chord_size_nested_group_chain_group_mid_single`, and similar).

I now added tests for these two cases and rebased on latest `master`. You can verify that only this changed since @thedrow 's review with `diff <(git -P diff 6fb4f9f6 e538d07c) <(git -P diff cbdebaec 505505881)`.

auvipy commented:
I think we can include this in a point release, if not that have to hold it as there is a 5.2.2 release pending tomorrow. any aditional docs change is not needed right?

auvipy reviewed (approved):
looks good. just want to wait some more time before merge

treszkai commented:
Thanks! No docs change necessary, as the code is logically equivalent to the previous version.

> On 17.12.2021, at 06:18, Asif Saif Uddin ***@***.***> wrote:
> 
> ﻿
> I think we can include this in a point release, if not that have to hold it as there is a 5.2.2 release pending tomorrow. any aditional docs change is not needed right?
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. 
> You are receiving this because you were mentioned.

auvipy commented:
then we can include this to next point release! thanks a lot for patience!
